`procurement.move_dest_id` is not necessary to add  fiscal_position map to procurement.rule.

ajalbut opened the pull request.

mileo commented:
André,

Sem o destinatário a regra de posição fiscal não sera calculada corretamente. A categoria pode até ser carregada mas a posição fiscal não estará correta.

mileo commented:
:-1:

ajalbut commented:
@mileo 
O campo `procurement.move_dest_id` não diz respeito ao destinatário da regra, mas sim ao `stock.move` que teria gerado o `procurement.order` por encadeamento. Mesmo em casos em que não há encadeamento (`move_dest_id` vazio), o destinatário deve ser associado, mas de forma que independe do referido campo.

Seu `res.partner` deve ser buscado de uma das seguintes formas:
1. diretamente na `procurement.rule` - `procurement.rule_id.partner_address_id`
2. no `procurement.group` - `procurement.group_id.partner_id`
3. no `stock.warehouse` de destino do `procurement.order` - `procurement.warehouse_id.company_id.partner_id`

Acredito que a alternativa 3 seja a mais robusta, uma vez que todos os campos encadeados são obrigatórios, o que elimina a necessidade de buscar por 1 ou 2, como atualmente está.

leorochael commented:
> Sem o destinatário a regra de posição fiscal não sera calculada corretamente. A categoria pode até ser carregada mas a posição fiscal não estará correta.

@mileo, isso que o @ajalbut falou é muito importante. `move_dest_id` não tem absolutamente nada a ver com o destinatário do procurement. É um m2o pra `stock.move`, supostamente o `stock.move` que deu origem ao procurement (e.g. quando um revendedor vende algo que não tem em estoque e precisa comprar. a venda é o `move_dest_id`, cujo `res.partner` é o cliente final, mas o `res.partner` que vc precisa para a posição fiscal do procurement é o revendedor, não o cliente final).

Um procurement sem `move_dest_id` é perfeitamente válido, e como o @ajalbut falou, não só não é necessário para localizar a posição fiscal pelo destinatário, como se você usá-lo para isso, você vai pegar o `res.partner` errado, e possivelmente a posição fiscal errada também.

leorochael commented:
@ajalbut,

> Seu res.partner deve ser buscado de uma das seguintes formas:
> 1. diretamente na procurement.rule - procurement.rule_id.partner_address_id
> 2. no procurement.group - procurement.group_id.partner_id
> 3. no stock.warehouse de destino do procurement.order - procurement.warehouse_id.company_id.partner_id
> 
> Acredito que a alternativa 3 seja a mais robusta, uma vez que todos os campos encadeados são obrigatórios, o que elimina a necessidade de buscar por 1 ou 2, como atualmente está.

Legal a sua análise. Pelo que eu entendi do código, o partner [atualmente é buscado apenas nas opções 1. e 2.](https://github.com/YouByPiccadilly/l10n-brazil/blob/c0a43ff0609ab5ca534f11cebb533c7df833b435/l10n_br_stock_account/models/procurement.py#L17) (ou seja, já não está seguindo a lógica que o @mileo [mencionou acima](#issuecomment-261453446)):

```
        partner = (
            procurement.rule_id.partner_address_id or (
                procurement.group_id and procurement.group_id.partner_id)
        )
```

Acredito que o correto seja modificar o código acima para buscar a primeira das três opções que você mencionou, na ordem que vc mencionou.

PS: Tem um erro de lógica no código acima, se `procurement.rule_id.partner_address_id` for falso e `procurement.group_id` também, `partner` vai ser um model `procurement.group` não navegado. O que dá um erro mais [abaixo no código ao tentar obter o atributo `partner.state_id`](https://github.com/YouByPiccadilly/l10n-brazil/blob/c0a43ff0609ab5ca534f11cebb533c7df833b435/l10n_br_stock_account/models/procurement.py#L41):

```
        product_fc_id = obj_fp_rule.with_context(
            ctx).product_fiscal_category_map(
                kwargs.get('product_id'),
                kwargs.get('fiscal_category_id'),
                partner.state_id.id)
```

Commit:
fixup! [FIX] Fix procurement rule company_id

`procurement.move_dest_id` is not necessary to add
fiscal_position map to procurement.rule.

`partner` was already located by
openerp.addons.stock.procurement:procurement_order

leorochael commented:
@mileo investigando aqui, chegamos à conclusão que a superclasse de `procurement.order` [já calcula o `partner` corretamente](https://github.com/odoo/odoo/blob/05de60cbf91e62ca03eee8fe91c0856dbd687941/addons/stock/procurement.py#L177).

Com esse commit novo, o @ajalbut não só corrige o bug de depender de `move_dest_id`, como também o cógido de obter o `partner` que estava bugado (estava faltando um `or False` o que causava a variável `partner`, em alguns casos, conter o recordset não navegado do modelo `procurement.group`).

Para mim este PR agora é :+1:

renatonlima reviewed (approved):
:+1:

mileo commented:
:+1:

leorochael commented:
@mileo, vc já é PSC. Pode mesclar este PR, please?

mileo commented:
@leorochael Not yet! But we need 3 approvals.

leorochael commented:
@mileo there. Three approvals. Only one by a maintainer, but this shouldn't be a problem.

The real problem is: why are you not a maintainer! I thought you had been approved as PSC!

@renatonlima how can you possibly justify @mileo not being a maintainer yet?

renatonlima commented:
Hello @mileo and @leorochael, 

Since 4th november I sent a email to OCA board to create the new satellites repos and add Mileo to core repo, I don't have the keys to make it. In last two weeks I was very busy and don't follow this request, I will check it.
